streams/src/main/java/org/apache/kafka/streams/processor/internals/RecordCollectorImpl.java

```diff
@@ -98,10 +108,17 @@ public void onCompletion(RecordMetadata metadata, Exception exception) {
         }
     }
 
+    private void checkForException() {
+        if (sendException != null) {
+            throw new StreamsException(String.format("%s exception caught when producing", logPrefix), sendException);
+        }
+    }
+
     @Override
     public void flush() {
         log.debug("{} Flushing producer", logPrefix);
         this.producer.flush();
+        checkForException();
```

Comment:
This is not introduced in this PR and we can tackle it in a separate PR if my reasoning is correct: in https://github.com/apache/kafka/pull/1970 we did this ordering:

```
1. Commit tasks BUT only commit their consumers, do not flush states or flush producers.
2. Close tasks.
3. Flush stores: stateMgr.flush(processorContext).
4. Flush producers.
5. Close state manager: closeAllStateManagers().
6. Close producer / consumer / restore consumer.
```

I do not remember is there is any reasons to do step 1 first, but it definitely void the at-least-once guarantees if we have a failure after step 1 before any of other steps right?